README.md
# swipeSlide for Zepto/jQuery Plugin

## 简介

移动端（基于Zepto/jQuery）的轮播插件：<http://ons.me/500.html>

## 示例

![扫一扫](website.png)
[普通轮播DEMO链接](http://ximan.github.io/swipeSlide/index.html)

![扫一扫](website-pic.png)
[全屏大图DEMO链接](http://ximan.github.io/swipeSlide/full-screen-pic.html)

![扫一扫](website-text.png)
[全屏文字DEMO链接](http://ximan.github.io/swipeSlide/full-screen-text.html)

![扫一扫](website-switch.png)
[快速切换DEMO链接](http://ximan.github.io/swipeSlide/index-switch.html)

![扫一扫](website-comment.png)
[点小图出大图DEMO链接](http://ximan.github.io/swipeSlide/comment-thumbnails.html)

## 依赖

Zepto 或者 jQuery

## 使用方法

````
$('.element').swipeSlide({
    // 参数
});
````

## 参数列表

|       参数        |   说明   |  默认值 |      可填值     |
|------------------|----------|--------|----------------|
| ul               | 父dom    | ul     | .element的子dom |
| li               | 子dom    | li     | ul的子dom       |
| index            | 轮播初始值 | 0     | 数字       |
| continuousScroll | 连续滚动   | false | true和false |
| autoSwipe        | 自动切换   | true  | true和false |
| speed            | 切换速度   | 4000  | 数字        |
| axisX            | X轴滚动   | true   | true和false |
| transitionType   | 过渡类型   | ease  | linear/ease/ease-in/ease-out/ease-in-out/cubic-bezier |
| lazyLoad         | 图片懒加载 | false | true和false |
| firstCallback    | 页面加载回调| 空    | function(i,sum,me){}（i为索引值，sum为总和，me为自己） |
| callback         | 每次滚动回调| 空    | function(i,sum,me){}（参数同上） |

## API

暴露一些功能，可以让swipeSlide更灵活的使用

`goTo(num)` 指定轮播，详见[DEMO4代码](index-switch.html)

## 最新版本

### 3.4.4(160909)

* 修复自动切换但连不续滚动时，手动操作首屏向前滑动或者末屏向后滑动，不自动切换bug

[所有更新日志](Changelog.md)

## 缺点

只能固定高度或者成比例宽度，无法自适应高度。

Changelog.md
### 3.4.4(160909)

* 修复自动切换但连不续滚动时，手动操作首屏向前滑动或者末屏向后滑动，不自动切换bug

### 3.4.3(160615)

* 修复点小图出大图DEMO，点击第4张图，滑动到第2张图无法懒加载bug

### 3.4.2(150930)

* 修复索引值为0时无法回调bug

### 3.4.1(150902)

* 修复不连续滚动时懒加载bug

### 3.4.0(150820)

* 支持多个轮播相同class调用
* 修改callback仅滚动时才回调（重要）
* 增加firstCallback页面加载回调
* 所有回调增加第三个参数me
* 修复点击触发callback回调bug

### 3.3.1(150427)

* 修复只有1个轮播时不执行lazyLoad和callback

### 3.3.0(150424)

* 增加轮播初始值index参数
* 修复点击触发轮播滚动

### 3.2.1(150331)

* 优化滑动时禁止另外一个方向的滑动事件

### 3.2.0(150322)

* 增加回调callback方法sum参数
* 修复连续滚动时只滚动一轮bug

### 3.1.0(150313)

* 支持指定索引值方法、前一屏、后一屏轮播
* 修复一屏内多图懒加载bug

### 3.0.0(150227)

* 重构js
* 支持resize
* 修改回调callback写法（废弃掉原来的callback写法）

### 2.2.1(150130)

* 优化懒加载时js阻塞导致轮播图片无法显示

### 2.2(150112)

* 修复Chrome模拟器touch变鼠标后js报错
* 修复只有1个轮播时禁止自动切换

### 2.1(150105)

* 支持Windows Phone

### 2.0(141007)

* 支持选择连续滚动
* 支持选择滚动轴
* 增加过渡效果（cubic-bezier）

### 1.1(140930)

* 支持Zepto精简版或jQuery 2.x库，去掉Zepto Touch模块依赖
* 增加图片跟手滑动
* 改善部分Android浏览器滑动卡顿

### 1.0(140905)

* 支持选择自动或手动切换
* 支持改变切换速度
* 支持图片懒加载

js/swipeSlide.js
/**
 * swipeSlide
 * http://ons.me/500.html
 * 西门
 * 3.4.4(160909)
 */
;(function(win,$){
    'use strict';

    // 判断IE
    var browser = {
        ie10 : win.navigator.msPointerEnabled,
        ie11 : win.navigator.pointerEnabled
    };
    // 触摸事件
    var events = ['touchstart','touchmove','touchend'];
    // 检测
    var support = {
        // 触摸
        touch : (win.Modernizr && Modernizr.touch === true) || (function () {
            return !!(('ontouchstart' in win) || win.DocumentTouch && document instanceof DocumentTouch);
        })()
    };

    // 判断浏览器
    if (browser.ie10) events = ['MSPointerDown', 'MSPointerMove', 'MSPointerUp'];
    if (browser.ie11) events = ['pointerdown', 'pointermove', 'pointerup'];

    // 触摸赋值
    var touchEvents = {
        touchStart : events[0],
        touchMove : events[1],
        touchEnd : events[2]
    };

    // 绑定swipeSlide
    $.fn.swipeSlide = function(options){
        var list = [];
        this.each(function(i, me){
            list.push(new sS(me, options));
        });
        return list;
    };
    var sS = function(element, options){
        var me = this;
        me.$el = $(element);
        me._distance = 50;
        me.allowSlideClick = true;
        me.init(options);
    };

    // 初始化
    sS.prototype.init = function(options){
        var me = this;
        me.opts = $.extend({}, {
            ul : me.$el.children('ul'),             // 父dom
            li : me.$el.children().children('li'),  // 子dom
            index : 0,                              // 轮播初始值
            continuousScroll : false,               // 连续滚动
            autoSwipe : true,                       // 自动切换
            speed : 4000,                           // 切换速度
            axisX : true,                           // X轴
            transitionType : 'ease',                // 过渡类型
            lazyLoad : false,                       // 图片懒加载
            firstCallback : function(){},           // 页面加载回调
            callback : function(){}                 // 每次滚动回调
        }, options);
        me._index = me.opts.index;
        // 轮播数量
        me._liLength = me.opts.li.length;
        me.isScrolling;

        // 回调
        me.opts.firstCallback(me._index,me._liLength,me.$el);

        // 如果轮播小于等于1个，跳出
        if(me._liLength <= 1){
            if(me.opts.lazyLoad) fnLazyLoad(me, 0);
            return false;
        }

        // 连续滚动，复制dom
        if(me.opts.continuousScroll) me.opts.ul.prepend(me.opts.li.last().clone()).append(me.opts.li.first().clone());

        // 懒加载图片
        if(me.opts.lazyLoad){
            fnLazyLoad(me, me._index);
            if(me.opts.continuousScroll){
                fnLazyLoad(me, me._index+1);
                fnLazyLoad(me, me._index+2);
                // 如果是第一屏
                if(me._index == 0){
                    fnLazyLoad(me, me._liLength);
                // 如果是最后一屏
                }else if(me._index+1 == me._liLength){
                    fnLazyLoad(me, 1);
                }
            }else{
                // 如果不连续滚动
                if(me._index == 0){
                    fnLazyLoad(me, me._index+1);
                }else if(me._index+1 == me._liLength){
                    fnLazyLoad(me, me._index-1);
                }else{
                    fnLazyLoad(me, me._index+1);
                    fnLazyLoad(me, me._index-1);
                }
            }
        }

        // 轮播的宽度
        fnGetSlideDistance();

        if(browser.ie10 || browser.ie11){
            // IE触控
            var action = '';
            if(me.opts.axisX){
                action = 'pan-y';
            }else{
                action = 'none';
            }
            me.$el.css({'-ms-touch-action':action,'touch-action':action});

            // 解决IE滑动触发click
            me.$el.on('click',function(){
                return me.allowSlideClick;
            });
        }

        // 调用轮播
        fnAutoSlide(me);

        // 绑定触摸
        me.$el.on(touchEvents.touchStart,function(e){
            fnTouches(e);
            fnTouchstart(e, me);
        });
        me.$el.on(touchEvents.touchMove,function(e){
            fnTouches(e);
            fnTouchmove(e, me);
        });
        me.$el.on(touchEvents.touchEnd,function(){
            fnTouchend(me);
        });
        me.opts.ul.on('webkitTransitionEnd MSTransitionEnd transitionend',function(){
            fnAutoSlide(me);
        });

        // 横竖屏、窗口调整
        $(win).on('onorientationchange' in win ? 'orientationchange' : 'resize',function(){
            clearTimeout(me.timer);
            me.timer = setTimeout(fnGetSlideDistance,150);
        });

        // 获取轮播宽度
        function fnGetSlideDistance(){
            var $li = me.opts.ul.children();
            me._slideDistance = me.opts.axisX ? me.opts.ul.width() : me.opts.ul.height();
            // 定位
            fnTransition(me, me.opts.ul, 0);
            fnTranslate(me, me.opts.ul, -me._slideDistance*me._index);
            fnTransition(me, $li, 0);
            var num = me.opts.continuousScroll ? -1 : 0;
            $li.each(function(i){
                fnTranslate(me, $(this), me._slideDistance*(i+num));
            });
        }
    };

    // css过渡
    function fnTransition(me, dom, num){
        dom.css({
            '-webkit-transition':'all '+num+'s '+me.opts.transitionType,
            'transition':'all '+num+'s '+me.opts.transitionType
        });
    }

    // css位移
    function fnTranslate(me, dom, distance){
        var result = me.opts.axisX ? distance+'px,0,0' : '0,'+distance+'px,0';
        dom.css({
            '-webkit-transform':'translate3d('+result+')',
            'transform':'translate3d('+result+')'
        });
    }

    // 懒加载图片
    function fnLazyLoad(me, index){
        var $li = me.opts.ul.children();
        var $thisImg = $li.eq(index).find('[data-src]');
        if($thisImg){
            $thisImg.each(function(i){
                var $this = $(this);
                if($this.is('img')){
                    $this.attr('src',$this.data('src'));
                    $this.removeAttr('data-src');
                }else{
                    $this.css({'background-image':'url('+$this.data('src')+')'});
                    $this.removeAttr('data-src');
                }
            });
        }
    }

    // touches
    function fnTouches(e){
        if(support.touch && !e.touches){
            e.touches = e.originalEvent.touches;
        }
    }

    // touchstart
    function fnTouchstart(e, me){
        me.isScrolling = undefined;
        me._moveDistance = me._moveDistanceIE = 0;
        // 按下时的坐标
        me._startX = support.touch ? e.touches[0].pageX : (e.pageX || e.clientX);
        me._startY = support.touch ? e.touches[0].pageY : (e.pageY || e.clientY);
    }

    // touchmove
    function fnTouchmove(e, me){
        // 如果自动切换，move的时候清除autoSlide自动轮播方法
        if(me.opts.autoSwipe) fnStopSlide(me);
        me.allowSlideClick = false;
        // 触摸时的坐标
        me._curX = support.touch ? e.touches[0].pageX : (e.pageX || e.clientX);
        me._curY = support.touch ? e.touches[0].pageY : (e.pageY || e.clientY);
        // 触摸时的距离
        me._moveX = me._curX - me._startX;
        me._moveY = me._curY - me._startY;
        // 优化触摸禁止事件
        if(typeof me.isScrolling == 'undefined'){
            if(me.opts.axisX){
                me.isScrolling = !!(Math.abs(me._moveX) >= Math.abs(me._moveY));
            }else{
                me.isScrolling = !!(Math.abs(me._moveY) >= Math.abs(me._moveX));
            }
        }

        // 距离
        if(me.isScrolling){
            if (e.preventDefault) e.preventDefault();
            else e.returnValue = false;
            // 触摸时跟手
            fnTransition(me, me.opts.ul, 0);
            me._moveDistance = me._moveDistanceIE = me.opts.axisX ? me._moveX : me._moveY;
        }
        if(!me.opts.continuousScroll){
            // 如果是第一屏，并且往下滚动，就不让滚动 || 如果是最后一屏，并且往上滚动，就不让滚动
            if(me._index == 0 && me._moveDistance > 0 || (me._index + 1) >= me._liLength && me._moveDistance < 0){
                me._moveDistance = 0;
            }
        }
        // 触摸时跟手滚动
        fnTranslate(me, me.opts.ul, -(me._slideDistance * me._index - me._moveDistance));
    }

    // touchend
    function fnTouchend(me){
        if(me.opts.autoSwipe){
            fnAutoSlide(me);
        }

        // 解决IE滑动触发click
        if(browser.ie10 || browser.ie11){
            if(Math.abs(me._moveDistanceIE) < 5){
                me.allowSlideClick = true;
            }
            setTimeout(function(){
                me.allowSlideClick = true;
            },100);
        }

        // 距离小
        if(Math.abs(me._moveDistance) <= me._distance){
            fnSlide(me, '', '.3');
        // 距离大
        }else{
            // 手指触摸上一屏滚动
            if(me._moveDistance > me._distance){
                fnSlide(me, 'prev', '.3');
            // 手指触摸下一屏滚动
            }else if(Math.abs(me._moveDistance) > me._distance){
                fnSlide(me, 'next', '.3');
            }
        }
        me._moveDistance = me._moveDistanceIE = 0;
    }

    // 自动轮播
    function fnAutoSlide(me){
        if(me.opts.autoSwipe){
            fnStopSlide(me);
            me.autoSlide = setInterval(function(){
                fnSlide(me, 'next', '.3');
            },me.opts.speed);
        }
    }

    // 停止轮播
    function fnStopSlide(me){
        clearInterval(me.autoSlide);
    }

    // 指定轮播
    sS.prototype.goTo = function(i){
        var me = this;
        fnSlide(me, i, '.3');
    };

    // 轮播方法
    function fnSlide(me, go, num){
        // 判断方向
        if(typeof go === 'number'){
            me._index = go;
            // 加载当前屏、前一屏、后一屏
            if(me.opts.lazyLoad){
                // 因为连续滚动，复制dom，所以要多加1
                if(me.opts.continuousScroll){
                    fnLazyLoad(me, me._index);
                    fnLazyLoad(me, me._index+1);
                    fnLazyLoad(me, me._index+2);
                }else{
                    fnLazyLoad(me, me._index-1);
                    fnLazyLoad(me, me._index);
                    fnLazyLoad(me, me._index+1);
                }
            }
        }else if(go == 'next'){
            me._index++;
            if(me.opts.lazyLoad){
                if(me.opts.continuousScroll){
                    fnLazyLoad(me, me._index+2);
                    // 滑到最后一屏
                    if(me._index+1 == me._liLength){
                        fnLazyLoad(me, 1);
                    // 最后一屏，继续往后滑动
                    }else if(me._index == me._liLength){
                        fnLazyLoad(me, 0);
                        fnLazyLoad(me, 2);
                    }
                }else{
                    fnLazyLoad(me, me._index+1);
                }
            }
        }else if(go == 'prev'){
            me._index--;
            if(me.opts.lazyLoad){
                if(me.opts.continuousScroll){
                    fnLazyLoad(me, me._index);
                    // 滑到第一屏
                    if(me._index == 0){
                        fnLazyLoad(me, me._liLength);

                    // 第一屏，继续往前滑动
                    }else if(me._index < 0){
                        fnLazyLoad(me, me._liLength-1);
                    }
                }else{
                    fnLazyLoad(me, me._index-1);
                }
            }
        }
        // 如果是连续滚动
        if(me.opts.continuousScroll){
            if(me._index >= me._liLength){
                fnScroll(me, num);
                me._index = 0;
                setTimeout(function(){
                    fnScroll(me, 0);
                    fnAutoSlide(me);
                //    return;
                },300);
            }else if(me._index < 0){
                fnScroll(me, num);
                me._index = me._liLength-1;
                setTimeout(function(){
                    fnScroll(me, 0);
                //   return;
                    fnAutoSlide(me);
                },300);
            }else{
                fnScroll(me, num);
            }
        }else{
            if(me._index >= me._liLength){
                me._index = 0;
            }else if(me._index < 0){
                me._index = me._liLength-1;
            }
            fnScroll(me, num);
        }
        // 如果第二个参数为空，就不回调
        if(arguments[1] !== ''){
            me.opts.callback(me._index,me._liLength,me.$el);
        }
    }

    // 轮播动作
    function fnScroll(me, num){
        fnTransition(me, me.opts.ul, num);
        fnTranslate(me, me.opts.ul, -me._index*me._slideDistance);
    }

})(window, window.Zepto || window.jQuery);
